### ISSUE TITLE
[Feature Request] Add customgraph module

### ISSUE THREAD

**Issue by otaj** [Alexays/Waybar#2684]
Hi, thank you for the awesome bar!

I was wondering, would it be possible to add a module creating a graph such as for example [CPUGraph from QTile](https://docs.qtile.org/en/stable/manual/ref/widgets.html#cpugraph) or [Graph from eww](https://elkowar.github.io/eww/widgets.html#graph). Ideally of course with being able to feed it custom data through continuous script similarly as custom module works.

I can now simulate it with the help of [Sparks font](https://github.com/aftertheflood/sparks), but it's a bit janky and it doesn't support all possible styling option.

I can help out a bit, however my C++ is very rusty so I wouldn't be the best help :)

Thank you!

**Comment by vincentaxhe** [comment]
that would be nice

**Comment by Gazer** [comment]
Hello, I've started to experiment to add support for real graphs (for now line only), this is how the CPU usage looks like:

<img width="223" height="48" alt="Image" src="https://github.com/user-attachments/assets/e1bde7ab-7abb-485c-84ae-f45e7e20ce0b" />

You can check the code at https://github.com/Gazer/Waybar/tree/feat/basic-cpu-graph

**Comment by imiric** [comment]
@Gazer That looks great! Would you be willing to create a PR for it?

And, more importantly, would @Alexays be interested in merging it?

**Comment by Gazer** [comment]
@imiric the idea is to end with a PR, right now I´m exploring to add a custom-graph similar to the custom module so you can use an external script to draw a graph and I want to add two more types: gauge and bars. E.g. for temperature a gauge has more sense a disk usage with bars or something like that.

**Comment by vincentaxhe** [comment]
@Gazer I'm interested in your work, the new module and how to configure it.

**Comment by Yilmaz4** [comment]
@Gazer any progress on this?

**Comment by Gazer** [comment]
I'm doing the final testing and polish but I may be able to do a PR to get feedback next week. I've working three graph types with native or custom modules. 

This config:
```
"custom-graph/memory": {
    "interval": 60,
    "graph_type": "gauge",
    "width": 52,
    "exec": "/home/gazer/mem.sh",
    "signal": 8,
    "return-type": "json"
  },
  "custom-graph/disk": {
    "interval": 2,
    "graph_type": "bar",
    "width": 10,
    "exec": "/home/gazer/disk.sh",
    "signal": 8,
    "return-type": "json"
  },
  cpu_graph": {
    "interval": 2,
    "graph_type": "line",
    "y_offset": 4,
    "width": 50,
    "on-click": "$TERMINAL -e btop"
  },
```
creates

<img width="250" height="36" alt="Image" src="https://github.com/user-attachments/assets/37a2e5ea-dd68-45fd-8f15-d6fca94fa74a" />

Missing things:
* man pages
* gauge and bar customization (right now the colors are hard-coded)

**Comment by Gazer** [comment]
PR created, I've drop the idea of the rainbown colors since do not match the style. Instead played with the alpha to create a shades of the main color

<img width="248" height="46" alt="Image" src="https://github.com/user-attachments/assets/ff189fb9-e3ef-454b-b8ff-580c6e892134" />
<img width="238" height="51" alt="Image" src="https://github.com/user-attachments/assets/a21d8dda-8ff8-4be6-844c-09649abb0852" />
<img width="264" height="69" alt="Image" src="https://github.com/user-attachments/assets/dc41a55d-e1da-41f5-b41e-f8376771db2d" />

**Comment by vincentaxhe** [comment]
@Gazer did the graph scale automatically,for example custom a network traffic line graph, when traffic change from 10k - 1M to 1M - 10M?

**Comment by Gazer** [comment]
Hi , no, this initial version just allow you to render values from 0 to 100, so the scale change is not implemented. It should work mostly for any value that can be mapped to a percentage for now since that's what was available in the custom json format.

A native network module may do some work to support re-scaling by recalculating all the values in the vector but I did not take the time to try it since I focused on the disk, memory and cpu.

**Comment by vincentaxhe** [comment]
@Gazer graph should implement something to indicate different scale, line color changed or vertical marks or fill density, so script can pass more data to scale.

**Comment by Gazer** [comment]
@vincentaxhe I agree, but since this is an initial implementation I rather prefer to keep it simple for now. I use this [gist](https://gist.github.com/Gazer/02c1d6d34b65278b056d2d79aa6c7f3c) to monitor network with autocale and work decently:

<img width="165" height="39" alt="Image" src="https://github.com/user-attachments/assets/5c9bbd81-cc7d-47e0-968f-5755dd37b867" />

```json
  "custom-graph/net-up": {
    "exec": "$HOME/netp.sh up",
    "width": 5,
    "graph_type": "bar",
    "return-type": "json"
  },
  "custom-graph/net-down": {
    "exec": "$HOME/netp.sh down",
    "width": 5,
    "graph_type": "bar",
    "return-type": "json"
  }
```
and the style:

```css
#custom-graph-net-up {
    color: #fa7070;
}

#custom-graph-net-down {
    color: #0080c0;
}
```

**Comment by Alexays** [comment]
Closing — this is implemented on current master and will be in the next release. Please reopen if it still occurs after updating. 🙏

**Comment by vincentaxhe** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。
